python_modules/dagster/dagster_tests/core_tests/resource_tests/test_without_resources.py

```diff
@@ -0,0 +1,198 @@
+    assert transformed_asset.required_resource_keys == {"the_key"}
+
+
+def test_source_asset():
+    foo_resource = ResourceDefinition.hardcoded_resource("foo")
+
+    @io_manager(required_resource_keys={"foo"})
+    def the_io_manager():
+        pass
+
+    the_asset = SourceAsset(
+        key=AssetKey("the_asset"),
+        io_manager_def=the_io_manager,
+        resource_defs={"foo": foo_resource},
+    )
+    assert the_asset.resource_defs == {
+        io_manager_key_for_asset_key(AssetKey("the_asset")): the_io_manager,
+        "foo": foo_resource,
+    }
+    transformed_asset = without_resources([the_asset])[0]
+    assert transformed_asset.resource_defs == {}
+    assert transformed_asset.get_io_manager_key() == DEFAULT_IO_MANAGER_KEY
+
+
+def test_source_asset_defines_io_manager_key():
+    @io_manager
+    def the_manager():
+        pass
+
+    the_asset = SourceAsset(
+        key=AssetKey("the_asset"), io_manager_key="the_key", resource_defs={"the_key": the_manager}
+    )
+    assert the_asset.get_io_manager_key() == "the_key"
+    transformed_asset = without_resources([the_asset])[0]
+    assert transformed_asset.get_io_manager_key() == "the_key"
+    assert transformed_asset.resource_defs == {}
+
+
+def test_io_manager_key_overrides_multiple_outputs():
+    # Simulate setting system-provided io manager key for multiple outs. Once https://github.com/dagster-io/dagster/issues/8650 lands, this should be replaced with using @multi_asset with AssetOut.
+
+    asset_key_a = AssetKey("a")
+    asset_key_b = AssetKey("b")
+
+    @op(
+        out={
+            "a": Out(io_manager_key=io_manager_key_for_asset_key(asset_key_a)),
+            "b": Out(io_manager_key=io_manager_key_for_asset_key(asset_key_b)),
+        }
+    )
+    def ma_op():
+        return 1
+
+    ma = AssetsDefinition(
+        node_def=ma_op,
+        keys_by_input_name={},
+        keys_by_output_name={"a": asset_key_a, "b": asset_key_b},
+    )
+
+    assert ma.required_resource_keys == {
+        io_manager_key_for_asset_key(asset_key_a),
+        io_manager_key_for_asset_key(asset_key_b),
+    }
+
+    ma_transformed = without_resources([ma])[0]
+    assert ma_transformed.required_resource_keys == {DEFAULT_IO_MANAGER_KEY}
+
+
+def test_without_resources_transitive_io_manager_deps():
+    @io_manager(required_resource_keys={"foo"})
+    def the_io_manager():
+        pass
+
+    @asset(
+        resource_defs={"foo": ResourceDefinition.hardcoded_resource("foo")},
+        io_manager_def=the_io_manager,
+    )
+    def the_asset():
+        pass
+
+    assert the_asset.required_resource_keys == {
+        "foo",
+        io_manager_key_for_asset_key(the_asset.asset_key),
+    }
+    transformed_asset = without_resources([the_asset])[0]
+    # Transitive resource dependencies are kept around, because there are is no way to specify resources as optional.
+    assert transformed_asset.required_resource_keys == {"foo", DEFAULT_IO_MANAGER_KEY}
```

Comment:
this is an unfortunate little tidbit. When a user provides a resource def directly to an asset, we can't know their intent. In order to make it possible to use provided resources directly within the op, we make them required on the op itself. This means that if I provide a resource required by my io manager, and then remove all resources, I still require the key transitively required by the io manager. If we had a notion of optional resources, we could default to making provided resource defs optional resource keys, so that it wouldn't be necessary to re-provide that value.

Reply:
ah I see, that makes sense